File: 24_10/15/P12983/solution.py
def solution(strs, t):
    # 활용할 변수 정리
    N = len(t)
    str_set = set(strs)
    min_counts = [20001 for _ in range(N)]
    
    # dp를 통해 최소 개수 연산
    for i in range(N):
        for j in range(i, min(N, i+5)):
            if (not t[i:j+1] in str_set): continue
            min_counts[j] = 1 if i == 0 else min(min_counts[j], min_counts[i-1] + 1)

    return -1 if min_counts[N-1] == 20001 else min_counts[N-1]  

File: 24_10/15/P12983/solution.js
function solution(strs, t) {
    // 활용할 변수 정리     
    const N = t.length;
    const strSet = new Set(strs);
    
    // dp를 통해 최소 단어 개수 반환
    const minCounts = new Array(N).fill(20001);
    for (let i = 0; i < N; i++) {
        const maxIdx = Math.min(N, i+5);
        for (let j = i; j < maxIdx; j++) {
            if (!strSet.has(t.slice(i, j+1))) continue;
            minCounts[j] = i === 0 ? 1 : Math.min(minCounts[i-1] + 1, minCounts[j]);
        }
    }

    return minCounts[N-1] === 20001 ? -1 : minCounts[N-1];
}